programs/compressed-pda/src/lib.rs

```diff
@@ -52,30 +85,312 @@ pub struct InstructionDataTransfer {
     proof_a: [u8; 32],
     proof_b: [u8; 64],
     proof_c: [u8; 32],
-    in_utxos: Vec<TokenInUtxo>,
     low_element_indexes: Vec<u16>,
     root_indexes: Vec<u64>,
     rpc_fee: Option<u64>,
-    out_utxo: Vec<TokenOutUtxo>,
+    in_utxo: Vec<Utxo>,
+    out_utxo: Vec<OutUtxo>,
 }
-use borsh::{BorshDeserialize, BorshSerialize};
 
+// TODO: parse utxos a more efficient way, since owner is sent multiple times this way
 #[derive(Debug)]
 #[account]
-pub struct TokenInUtxo {
+pub struct InstructionDataTransfer2 {
+    proof_a: [u8; 32],
+    proof_b: [u8; 64],
+    proof_c: [u8; 32],
+    low_element_indexes: Vec<u16>,
+    root_indexes: Vec<u64>,
+    rpc_fee: Option<u64>,
+    utxos: SerializedUtxos,
+}
+
+// there are two sources I can get the pubkey from the transaction object and the other account keys
+// the index starts with the accounts keys of the transaction object, if the index is larger than the length of the accounts keys
+// we access the pubkey array with our additiona pubkeys
+
+// we need a general macro that just derives a serializable struct from a struct that replaces every pubkey with a u8
+// the struct should be borsh serializable and deserializable
+// additionally the macro should derive a function that converts the serializable struct back to the original struct with the additional input
+// of a slice of account infos where it gets the pubkeys from
+// additionally the macro needs to derive a function that converts the original struct into the serializable struct and outputs the struct and the pubkeys
+#[derive(Debug, Clone, AnchorSerialize, AnchorDeserialize)]
+pub struct SerializedUtxos {
+    pub pubkey_array: Vec<Pubkey>,
+    pub u64_array: Vec<u64>,
+    pub in_utxo: Vec<InUtxoSerializable>,
+    pub out_utxo: Vec<OutUtxoSerializable>,
+}
+
+impl SerializedUtxos {
+    pub fn in_utxos_from_serialized_utxos(
+        &self,
+        accounts: &[Pubkey],
+        merkle_tree_accounts: &[Pubkey],
+    ) -> Vec<Utxo> {
+        let mut in_utxos = Vec::new();
+        for (i, in_utxo) in self.in_utxo.iter().enumerate() {
+            let owner = if (in_utxo.owner as usize) < accounts.len() {
+                accounts[in_utxo.owner as usize]
+            } else {
+                self.pubkey_array[in_utxo.owner.saturating_sub(accounts.len() as u8) as usize]
+            };
+            let lamports = self.u64_array[in_utxo.lamports as usize];
+            let data = in_utxo.data.as_ref().map(|data| {
+                data.tlv_from_serializable_tlv(
+                    [accounts, self.pubkey_array.as_slice()].concat().as_slice(),
+                )
+            });
+            let mut utxo = Utxo {
+                owner,
+                blinding: [0u8; 32],
+                lamports,
+                data,
+            };
+            utxo.update_blinding(merkle_tree_accounts[i].key(), in_utxo.leaf_index as usize)
+                .unwrap();
```

Comment:
```suggestion
                ?;
```

and  `Result<Vec<Utxo>>` as a type